### ISSUE TITLE
Commit messages should reference either a GitHub issue or a JIRA key

### ISSUE THREAD

**Issue by cameronfletcher** [LykkeCorp/guidelines#5]
Obviously.
Discuss.

**Comment by samodovdi** [comment]
maybe to implement pre-commit hook for validation purposes

**Comment by JoeStead** [comment]
I disagree with the commit message approach. Commit messages should include whatever they want, the PR should refer to the issue in question, and when merged into the target branch, the linked issue will be there (as the commit will be from the PR).

As for using pre-commit hooks, I've never seen them scale correctly, they're a client-side thing.

**Comment by jchannon** [comment]
I disagree, if a large PR is a set of commits then this will wont be feasible. The PR should link to the Issue though

**Comment by cameronfletcher** [comment]
The other issue here, with the PR approach, is that does not result in the Git repository containing all the history of the linked issues. That information would only exist in GitHub. I would prefer the Git repository to effectively be self-contained.

**Comment by JoeStead** [comment]
I don't understand your last point. 

The PRs when merged will result in a commit on the target branch which link to the issue in question, it's no different from using commit messages, but it means when working on a feature you don't have to use the issue in every message, it just become s noise imo

**Comment by jchannon** [comment]
For example
https://github.com/LykkeCorp/apollo/commit/d7b744672549e4dcc0d060e5e23ae22e5f152b57

**Comment by ozhelenkov** [comment]
It raises a question about merging strategy.

If we use a merge commit then as for me it makes sense to reference a task key in a branch.

As for JIRA / GitHub the question seems weird for me. If you use JIRA to track project tasks, it make sense to use JIRA keys. To provide updates for community raised issues you can also add additional github reference as for me. Either close an issue with a proper comment.
